src/main/java/software/amazon/awssdk/crt/auth/credentials/CognitoCredentialsProvider.java

```diff
@@ -0,0 +1,265 @@
+        super();
+
+        String endpoint = builder.getEndpoint();
+        String identity = builder.getIdentity();
+        if (endpoint == null || identity == null) {
+            throw new IllegalArgumentException("CognitoCredentialsProvider - endpoint and identity must not be null");
+        }
+
+        ClientBootstrap clientBootstrap = builder.getClientBootstrap();
+        if (clientBootstrap == null) {
+            clientBootstrap = ClientBootstrap.getOrCreateStaticDefault();
+        }
+
+        TlsContext tlsContext = builder.getTlsContext();
+        if (clientBootstrap == null || tlsContext == null) {
+            throw new IllegalArgumentException("CognitoCredentialsProvider - clientBootstrap and tlsContext must not be null");
+        }
+
+        int proxyConnectionType = 0;
+        long proxyTlsContextHandle = 0;
+        String proxyHost = null;
+        int proxyPort = 0;
+        int proxyAuthorizationType = 0;
+        String proxyAuthorizationUsername = null;
+        String proxyAuthorizationPassword = null;
+        HttpProxyOptions proxyOptions = builder.getHttpProxyOptions();
+        if (proxyOptions != null) {
+            proxyConnectionType = proxyOptions.getConnectionType().getValue();
+            TlsContext proxyTlsContext = proxyOptions.getTlsContext();
+            if (proxyTlsContext != null) {
+                proxyTlsContextHandle = proxyTlsContext.getNativeHandle();
+            }
+
+            proxyHost = proxyOptions.getHost();
+            proxyPort = proxyOptions.getPort();
+            proxyAuthorizationType = proxyOptions.getAuthorizationType().getValue();
+            proxyAuthorizationUsername = proxyOptions.getAuthorizationUsername();
+            proxyAuthorizationPassword = proxyOptions.getAuthorizationPassword();
+        }
+
+        long nativeHandle = cognitoCredentialsProviderNew(
+                this,
+                clientBootstrap.getNativeHandle(),
+                tlsContext.getNativeHandle(),
+                endpoint,
+                identity,
+                builder.getCustomRoleArn(),
+                marshalLoginsForJni(builder.getLogins()),
+                proxyConnectionType,
+                proxyHost != null ? proxyHost.getBytes(UTF8) : null,
+                proxyPort,
+                proxyTlsContextHandle,
+                proxyAuthorizationType,
+                proxyAuthorizationUsername != null ? proxyAuthorizationUsername.getBytes(UTF8) : null,
+                proxyAuthorizationPassword != null ? proxyAuthorizationPassword.getBytes(UTF8) : null);
+
+        acquireNativeHandle(nativeHandle);
+        addReferenceTo(clientBootstrap);
+        addReferenceTo(tlsContext);
+    }
+
+    private byte[] marshalLoginsForJni(ArrayList<CognitoLoginTokenPair> logins) {
+        int size = 0;
+
+        for (CognitoLoginTokenPair login : logins) {
+            size += login.identityProviderName.length + login.identityProviderToken.length + (BUFFER_INT_SIZE * 2);
```

Comment:
Can `identityProviderName` be null? If it's null, the exception will be raised here.

Reply:
I don't think name can be null but it might be possible that token might be in an unauthenticated setting.  So I'll add protections around both anyways.